### PR TITLE
8298848: C2: clone all of (CmpP (LoadKlass (AddP down at split if

### PR DESCRIPTION
**Pull request by rwestrel**
As suggested by Vladimir in:
https://github.com/openjdk/jdk/pull/11666

Thus extract one for the fixes as a separate PR. The bug as described
in the above PR is:

The crash occurs because a` (If (Bool (CmpP (LoadKlass ..))))`
only has a single projection. It lost the other projection because of
a `CheckCastPP` that becomes `top`. Initially the pattern is, in pseudo
code:

```
if (obj.klass == some_class) {
  obj = CheckCastPP#1(obj);
}
```

`obj` itself is a `CheckCastPP` that's pinned at a dominating if. That
dominating if goes through split through phi. The `LoadKlass` for the
pseudo code above also has control set to the dominating if being
transformed. This result in:

```
if (phi1 == some_class) {
  obj = CheckCastPP#1(phi2);
}
```

with` phi1 = (Phi (LoadKlass obj) (LoadKlass obj))` and phi2 = (Phi obj obj)
with `obj = (CheckCastPP#2 obj')`

`PhiNode::Ideal()` transforms `phi2` into a new `CheckCastPP`:
`(CheckCastPP#3 obj' obj') `with control set to the region right above
the if in the pseudo code above. There happens to be another
`CheckCastPP` at the same control which casts obj' to a narrower
type. So the new `CheckCastPP#3` is replaced by that one (because of
`ConstraintCastNode::dominating_cast()`) and pseudo code becomes:

```
if (phi1 == some_class) {
  obj = CheckCastPP#1(CheckCastPP#4(obj'));
}
```

and then:

```
if (phi1 == some_class) {
  obj = top;
}
```

because the types of the 2 `CheckCastPP`s conflict. That would be ok if:

`phi1 == some_class`

would constant fold. It would if the test was:

`if (CheckCastPP#4(obj').klass == some_klass) {
`
but because of split if, the `(CmpP (LoadKlass ..))` and the
`CheckCastPP#1` ended up with 2 different object inputs that then were
transformed differently. The fix I propose is to have split if clone the entire:

`(Bool (CmpP (LoadKlass (AddP ..))))`

down the same way `(Bool (CmpP ..))` is cloned down. After split if, the
pseudo code becomes:

```
if (phi.klass == some_class) {
  obj = CheckCastPP#1(phi);
}

```
The bug can't occur because the `CheckCastPP` and` (CmpP (LoadKlass ..))`
operate on the same phi input. The change in split_if.cpp implements
that.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8298848](https://bugs.openjdk.org/browse/JDK-8298848): C2: clone all of (CmpP (LoadKlass (AddP down at split if

### Reviewers
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**) ⚠️ Review applies to [e4b4c299](https://git.openjdk.org/jdk/pull/11689/files/e4b4c2996c6f685dfb6a296478558658b868de5e)
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**) ⚠️ Review applies to [31fc40f7](https://git.openjdk.org/jdk/pull/11689/files/31fc40f727e35cef07d9d80647bd8d7a48b688dc)
 * [Christian Hagedorn](https://openjdk.org/census#chagedorn) (@chhagedorn - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/11689/head:pull/11689` \
`$ git checkout pull/11689`

Update a local copy of the PR: \
`$ git checkout pull/11689` \
`$ git pull https://git.openjdk.org/jdk pull/11689/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 11689`

View PR using the GUI difftool: \
`$ git pr show -t 11689`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/11689.diff">https://git.openjdk.org/jdk/pull/11689.diff</a>

</details>

### CODE REVIEW

**Review by vnkozlov — COMMENTED** [review]
Looks good. I will test it.

**Review by vnkozlov — APPROVED** [review]
My tier1-4, xcomp, stress testing passed.

**Comment by vnkozlov** [comment]
You need second review.

**Comment by rwestrel** [comment]
Anyone else for this one? (#11666 depends on it and I prepared #11673 on top of #11666 to avoid merge conflicts so it indirectly depends on it too).

**Comment by rwestrel** [comment]
I added a `-XX:+IgnoreUnrecognizedVMOptions` to the test case so it runs on 32 bits (compressed oops not supported there)

**Review by TobiHartmann — APPROVED** [review]
Looks reasonable to me.

File: src/hotspot/share/opto/split_if.cpp
Comment (chhagedorn) [inline src/hotspot/share/opto/split_if.cpp:74]:
```suggestion
  if (!at_relevant_ctrl(n, blk1, blk2))
```

File: src/hotspot/share/opto/split_if.cpp
Comment (chhagedorn) [inline src/hotspot/share/opto/split_if.cpp:88]: I suggest to also rename this into something like `clone_cmp_loadklass_down` to follow the naming of `clone_cmp_down`.

File: test/hotspot/jtreg/compiler/types/TestCheckCastPPBecomesTOP.java
Comment (chhagedorn) [reply test/hotspot/jtreg/compiler/types/TestCheckCastPPBecomesTOP.java:2]: You should also update the copyright years in the other touched files.

**Review by chhagedorn — APPROVED** [review]
The fix looks reasonable to me, too.

**Review by chhagedorn — APPROVED** [review]
Thanks for doing the updates, looks good!

**Comment by rwestrel** [comment]
@vnkozlov @TobiHartmann @chhagedorn thanks for the reviews/testing.

**Pull request closed without merging**
